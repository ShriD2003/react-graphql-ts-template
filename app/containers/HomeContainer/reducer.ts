import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash-es/get';
import { Launch } from '@app/containers/HomeContainer';

export type HomeContainerState = {
  launchData: {
    launches?: Launch[];
  };
  launchListError: any;
  loading: boolean;
};

export const { Types: homeContainerTypes, Creators: homeContainerCreators } = createActions({
  requestGetLaunchList: { missionName: null, order: 'asc', page: 1 },
  successGetLaunchList: ['launchData'],
  failureGetLaunchList: ['launchListError']
});

export const initialState: HomeContainerState = {
  loading: false,
  launchData: {},
  launchListError: null
};

export interface HomeContainerAction extends Partial<Omit<HomeContainerState, 'loading'>> {
  type?: string;
  order?: string;
  page?: number;
  missionName?: string;
}

export const homeContainerReducer = (state = initialState, action: HomeContainerAction) =>
  produce(state, (draft) => {
    switch (action.type) {
      case homeContainerTypes.REQUEST_GET_LAUNCH_LIST:
        draft.loading = true;
        break;

      case homeContainerTypes.SUCCESS_GET_LAUNCH_LIST:
        draft.launchData = action.launchData!;
        draft.launchListError = null;
        draft.loading = false;
        break;
      case homeContainerTypes.FAILURE_GET_LAUNCH_LIST:
        draft.loading = false;
        draft.launchListError = draft.launchListError = get(action.launchListError, 'message', 'something_went_wrong');
        break;
    }
  });

export default homeContainerReducer;
