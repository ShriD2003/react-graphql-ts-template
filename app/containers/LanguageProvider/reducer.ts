/*
 *
 * LanguageProvider reducer
 *
 */
import { createActions } from 'reduxsauce';
import produce from 'immer';
import { DEFAULT_LOCALE } from '@app/i18n';

type LanguageProviderTypes = {
  CHANGE_LOCALE: 'CHANGE_LOCALE';
};

type LanguageProviderAction = {
  type?: keyof LanguageProviderTypes;
  locale?: string;
};

type LanguageProviderCreators = {
  changeLocale: (locale: 'en') => LanguageProviderAction;
};

export const { Types: languageProviderTypes, Creators: languageProviderActions } = createActions<
  LanguageProviderTypes,
  LanguageProviderCreators
>({
  changeLocale: ['locale']
});

export const initialState = {
  locale: DEFAULT_LOCALE
};

/* eslint-disable default-case, no-param-reassign */
export const languageProviderReducer = (state = initialState, action: LanguageProviderAction) =>
  produce(state, (draft) => {
    switch (action.type) {
      case languageProviderTypes.CHANGE_LOCALE:
        if (action.locale) {
          draft.locale = action.locale;
        }
        break;
    }
  });

export default languageProviderReducer;
