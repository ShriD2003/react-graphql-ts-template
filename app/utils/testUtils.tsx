import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
