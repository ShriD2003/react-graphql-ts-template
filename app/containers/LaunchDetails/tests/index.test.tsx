/**
 *
 * Tests for LaunchDetails
 *
 *
 */

import React from 'react';
import { renderProvider, timeout } from '@utils/testUtils';
// import { fireEvent } from '@testing-library/dom'
import { LaunchDetailsProps, LaunchDetailsTest as LaunchDetails, mapDispatchToProps } from '..';
import history from '@app/utils/history';
import { requestGetLaunch } from '../reducer';

describe('<LaunchDetails /> container tests', () => {
  let submitSpy: jest.Mock;
  let defaultProps: LaunchDetailsProps;

  beforeEach(() => {
    submitSpy = jest.fn();
    defaultProps = {
      loading: false,
      launch: null,
      dispatchLaunch: submitSpy
    };
  });
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<LaunchDetails {...defaultProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchLaunch if launchId param is present', async () => {
    history.location.pathname = '/1';
    renderProvider(<LaunchDetails {...defaultProps} />, { path: '/:launchId' });
    await timeout(500);
    expect(submitSpy).toBeCalledWith('1');
  });

  it('should mapDispatchToProps works as expected', () => {
    const dispatchSpy = jest.fn();
    const props = mapDispatchToProps(dispatchSpy);
    props.dispatchLaunch('1');
    expect(dispatchSpy).toBeCalledWith(requestGetLaunch('1'));
  });
});
