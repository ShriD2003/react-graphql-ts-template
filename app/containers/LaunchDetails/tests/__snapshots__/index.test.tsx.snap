// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchDetails /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="NotFoundPage__NotFoundContainer-ywwv48-1 hYGWmJ"
      >
        <p
          class="T__StyledText-gcrjip-0 kPLUDf"
          data-testid="details"
        >
          Oops you ran out of Oxygen!, dont worry, you can go back home and start exploring again!
        </p>
        <button
          class="NotFoundPage__CustomButton-ywwv48-2 inGbhQ"
          data-testid="back-button"
        >
          Go Back
        </button>
        <div
          class="ant-image"
        >
          <img
            class="ant-image-img NotFoundPage__NotFoundImage-ywwv48-0 eLTHHc"
            src="IMAGE_MOCK"
          />
        </div>
      </div>
      <p
        class="T__StyledText-gcrjip-0 cSxVIW"
        data-testid="default-message"
      />
    </div>
  </div>
</body>
`;
